Portable html: the css and javascript parts

oscarlevin opened the pull request:
This adds a new publication variable at `/publication/html/platform/@portable`, which when set to "yes" will result in the HTML using a CDN for all css and javascript, and not copy those files to the output.

This is set up to allow the CLI to set a string-param for `cli.version`, which will be added to the URL for the CDN (since we will keep the versions of the css and javascript versioned identically to the cli).

The documentation and schema for the publication file had outdated information about `platform`: they said that `@platform` was an attribute of `/publication/html`, rather than `platform` being a child element of `/publication/html`.  That is fixed in the guide and publication schema, adjacent to the new `@portable` attribute.

Happy to rename things as needed, but this does seem functional now.

rbeezer commented:
Appears a `pretext/pretext` build will use `latest`, right?  That's a good thing.  (haven't looked - is that documented?)

rbeezer commented:
This has a merge conflict.  Probably with #2403?

I thought I fixed it OK, but maybe I need an `ensure` call prior.  Otr is that #2404.  :-(  Anyway, rrebasoing, guessing at a fix, then building HTML gives:

```
Traceback (most recent call last):
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 855, in <module>
    main()
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 731, in main
    ptx.html(
  File "/home/rob/mathbook/mathbook/pretext/pretext.py", line 3864, in html
    if get_publisher_variable(xml, pub_file, stringparams, 'portable-html') != "yes":
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

oscarlevin commented:
Right, the correct fix would be to use `ensure` but that would require #2404.  Which is why I was waiting to resolve the merge conflict to see which PR would win.

rbeezer commented:
Bloddy hell.  ;-)  Well, I'm invested in #2404 and this.  So you do want you want and tell me what to do first?  ;-)

oscarlevin commented:
I just pushed a commit on #2404, but can't figure out the epub thing (I get other errors, so can't figure out how to test correctly).

I think though that this here is easier to fix and less controversial, so perhaps let me resolve the conflict and you can do this one first.

oscarlevin commented:
That should be correct. And it should be safe to squash just the last two commits

rbeezer commented:
Your last addition is a merge of master into solo-html.  When I view it online, I'm not sure what I am seeing.

Can you **rebase** your `solo-html` branch and force-push?

oscarlevin commented:
Done.  I tried using github's online merge conflict resolver, and thought it was just committing that one additional change.  Last time I do that!

rbeezer commented:
Still getting
```
Traceback (most recent call last):
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 855, in <module>
    main()
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 731, in main
    ptx.html(
  File "/home/rob/mathbook/mathbook/pretext/pretext.py", line 3864, in html
    if get_publisher_variable(xml, pub_file, stringparams, 'portable-html') != "yes":
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: get_publisher_variable() takes 2 positional arguments but 4 were given
```
I don't see `get_publisher_variable()` defined to take four arguments.

oscarlevin commented:
oooooh.  Okay, I thought it was just the merge conflict.  Went back and modified the functions to align with #2403.  Even tested it in both configurations now.  Third time lucky.

rbeezer commented:
>  Third time lucky.

`<audible-chuckle/>`

rbeezer commented:
Sample article has "custom" list markers.  Which go into their own CSS file.  Which I think is always non-empty (and so gets copied along).  Is this file being not-created?  Not sure whay this fails this way.  Will try digging deeper.
```
Traceback (most recent call last):
  File "/usr/lib/python3.11/shutil.py", line 825, in move
    os.rename(src, real_dst)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ptx-rwsmpkhn/ol-markers.css' -> '/tmp/ptx-rwsmpkhn/_static/pretext/css/ol-markers.css'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 855, in <module>
    main()
  File "/home/rob/mathbook/mathbook/pretext/pretext", line 731, in main
    ptx.html(
  File "/home/rob/mathbook/mathbook/pretext/pretext.py", line 3898, in html
    move_ol_marker_css(tmp_dir)
  File "/home/rob/mathbook/mathbook/pretext/pretext.py", line 3761, in move_ol_marker_css
    shutil.move(src, dest)
  File "/usr/lib/python3.11/shutil.py", line 845, in move
    copy_function(src, real_dst)
  File "/usr/lib/python3.11/shutil.py", line 436, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.11/shutil.py", line 258, in copyfile
    with open(dst, 'wb') as fdst:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ptx-rwsmpkhn/_static/pretext/css/ol-markers.css'
```

oscarlevin commented:
Huh.  I had not anticipated dynamically generated css.  Not sure how to deal with that, unless we write it to the html.

rbeezer commented:
> Not sure how to deal with that, unless we write it to the html.

I think @ascholerChemeketa had some ideas about how to do that.

oscarlevin commented:
Until a better solution presents itself, let's just not try to copy that file for non-portable builds.  

This means we are testing for the portable build three times in the html function.  I simplified this by creating a boolean `include_static_files` to make the code look prettier.  I'm not sure how important the order is (place runestone, then copy managed directories, then copy pretext js/css, then xsltproc, then move ol markes), but I'm afraid to mess with it right now.

Tested with the sample article in portable and non-portable modes.  Both build; portable defaults back to non-custom markers, but still looks okay.

rbeezer commented:
Code comment: `'Add a prefix for the cdn, which is empty unless the portable html variable is true`

Equals "latest", not "empty", no?

The Runestone version is not being read from the Runestone services file, so I see an empty string in `eBookConfig` as part of a diff before/after.  Similarly, I suspect variables like `$rs-js` are empty, so when tokenized, the looping (`for-each`) is empty, so no `link`/`script` are being written into the head.  Not clear to me just which **change** is provoking this.

Shouldn't we be pointing to the Runestone CDN?  Or does the Runestone Services JS/CSS live on `jsdelivr` and it needs to be pointed to?  Certain features, like a multiple choice problem are active without hosting on Runestone.

ascholerChemeketa commented:
The idea for generated extra CSS was that we can try to include it in the theme building process - have the `cssbuilder` script slurp up the ol-markers file and any future ones and bake them into the theme file. But that doesn't help here.

oscarlevin commented:
Okay, I think I fixed everything, including a missed xsl diagcess template which wasn't getting the cdn prefix.

Added a new _rs_* function specific for portable builds.  It might make sense to refactor all these functions at some point, but this was the simplest change for now.  Diffs of html come back as expected (although as I expect, I might be missing something again).

rbeezer commented:
No ill effects on non-portable versions that I can see, so this can move forward.

I would have liked to split out the script stuff from the HTML stuff, but pushed prematurely.

I see a `link` to the `ol-markers.css` file still.  File may be gone, but pointer is still there, I think.

I'd love to see "single-page" HTML as a real possibility for limited documents with all this infrastructure.  I know that is not identical to "portable".

*  Worksheets, `index.html`, and "standalone" pages for `interactive`, `video`, `audio` could all go away in output.  I suspect it could be easy to kill/neuter the templates making those things?
*  Interactives (especially the more custom ones) may need some associated code, similar to external images.
*  It'd be nice to eventually restore some of the documentation about single-page HTML: do this (e.g. chunk level 0), don't do this (e.g. interactives with custom JS), etc.

rbeezer commented:
I'm going to pick up the first three commits here, so I can attack a couple other PRs.

rbeezer commented:
Ignore that!  Temporarily lost in my own repo.....
